viscy/representation/evaluation/distance.py

```diff
@@ -21,84 +30,68 @@ def calculate_cosine_similarity_cell(embedding_dataset, fov_name, track_id):
 
     Returns
     -------
     dict[int, list[float]]
-        Dictionary mapping τ to list of displacements for all particles and initial times
+        Dictionary mapping time lag τ to list of squared displacements
     """
-    # Get unique tracks efficiently using pandas operations
+
     unique_tracks_df = (
         embedding_dataset[["fov_name", "track_id"]].to_dataframe().drop_duplicates()
     )
 
-    # Get data from dataset
-    fov_names = embedding_dataset["fov_name"].values
-    track_ids = embedding_dataset["track_id"].values
-    timepoints = embedding_dataset["t"].values
-    embeddings = embedding_dataset["features"].values
-
-    # Initialize results dictionary with empty lists
     displacement_per_tau = defaultdict(list)
 
-    # Process each track
     for fov_name, track_id in zip(
         unique_tracks_df["fov_name"], unique_tracks_df["track_id"]
     ):
-        # Get sorted track data
-        mask = (fov_names == fov_name) & (track_ids == track_id)
-        times = timepoints[mask]
-        track_embeddings = embeddings[mask]
+        # Filter data for this track
+        track_data = embedding_dataset.where(
+            (embedding_dataset["fov_name"] == fov_name)
+            & (embedding_dataset["track_id"] == track_id),
+            drop=True,
+        )
 
         # Sort by time
-        time_order = np.argsort(times)
-        times = times[time_order]
-        track_embeddings = track_embeddings[time_order]
-
-        # Process each time point
-        for t_idx, t in enumerate(times[:-1]):
-            current_embedding = track_embeddings[t_idx]
-
-            # Check all possible future time points
-            for future_idx, future_time in enumerate(
-                times[t_idx + 1 :], start=t_idx + 1
-            ):
-                tau = future_time - t
-                future_embedding = track_embeddings[future_idx]
-
-                if distance_metric in ["cosine"]:
-                    dot_product = np.dot(current_embedding, future_embedding)
-                    norms = np.linalg.norm(current_embedding) * np.linalg.norm(
-                        future_embedding
-                    )
-                    similarity = dot_product / norms
-                    displacement = similarity
-                else:  # Euclidean metrics
-                    diff_squared = np.sum((current_embedding - future_embedding) ** 2)
-                    displacement = diff_squared
-                displacement_per_tau[int(tau)].append(displacement)
+        time_order = np.argsort(track_data["t"].values)
+        times = track_data["t"].values[time_order]
+        track_embeddings = track_data["features"].values[time_order]
+
+        # Compute pairwise distance matrix
+        if distance_metric == "euclidean":
+            distance_matrix = pairwise_distance_matrix(
+                track_embeddings, metric="euclidean"
+            )
+            distance_matrix = distance_matrix**2  # Square for MSD
```

Comment:
Can you use the `seuclidean` metric directly?

https://github.com/scipy/scipy/blob/b1296b9b4393e251511fe8fdd3e58c22a1124899/scipy/spatial/distance.py#L2877-L2888

```
    4. ``Y = cdist(XA, XB, 'seuclidean', V=None)``

       Computes the standardized Euclidean distance. The standardized
       Euclidean distance between two n-vectors ``u`` and ``v`` is

       .. math::

          \\sqrt{\\sum {(u_i-v_i)^2 / V[x_i]}}.

       V is the variance vector; V[i] is the variance computed over all
       the i'th components of the points. If not passed, it is
       automatically computed.
```

Reply:
If the post-processing is removed, this branching logic becomes unnecessary.